docs/docs/module_guides/workflow/index.md

```diff
@@ -0,0 +1,203 @@
+        response = await self.llm.acomplete(prompt)
+        return JokeEvent(joke=str(response))
+
+    @step()
+    async def critique_joke(self, ev: JokeEvent) -> StopEvent:
+        joke = ev.joke
+
+        prompt = f"Give a thorough analysis and critique of the following joke: {joke}"
+        response = await self.llm.acomplete(prompt)
+        return StopEvent(result=str(response))
+
+
+w = JokeFlow(timeout=60, verbose=False)
+result = await w.run(topic="pirates")
+print(str(result))
+```
+
+There's a few moving pieces here, so let's go through this piece by piece.
+
+### Defining Workflow Events
+
+```python
+class JokeEvent(Event):
+    joke: str
+```
+
+Events are user-defined pydantic objects. You control the attributes and any other auxiliary methods. In this case, our workflow relies on a single user-defined event, the `JokeEvent`.
+
+### Setting up the Workflow Class
+
+```python
+class JokeFlow(Workflow):
+    llm = OpenAI(model="gpt-4o-mini")
+    ...
+```
+
+Our workflow is implemented by subclassing the `Workflow` class. For simplicity, we attached a static `OpenAI` llm instance.
+
+### Workflow Entry Points
+
+```python
+class JokeFlow(Workflow):
+    ...
+
+    @step()
+    async def generate_joke(self, ev: StartEvent) -> JokeEvent:
+        topic = ev.get("topic")
+
+        prompt = f"Write your best joke about {topic}."
+        response = await self.llm.acomplete(prompt)
+        return JokeEvent(joke=str(response))
+
+    ...
+```
+
+Here, we come to the entry-point of our workflow. While events are use-defined, there are two special-case events, the `StartEvent` and the `StopEvent`. Here, the `StartEvent` signifies where to send the initial workflow input. Since the input can be anything, its accessed safely using the `.get()` method.
+
+At this point, you may have noticed that we haven't explicitly told the workflow what events are handled by which steps. Instead, the `@step()` decorator is used to infer the input and output types of each step. Furthermore, these inferred input and output types are also used to verify for you that the workflow is valid before running!
+
+### Workflow Exit Points
+
+```python
+class JokeFlow(Workflow):
+    ...
+
+    @step()
+    async def critique_joke(self, ev: JokeEvent) -> StopEvent:
+        joke = ev.joke
+
+        prompt = f"Give a thorough analysis and critique of the following joke: {joke}"
+        response = await self.llm.acomplete(prompt)
+        return StopEvent(result=str(response))
+
+    ...
+```
+
+Here, we have our second, and last step, in the workflow. We know its the last step because the special `StopEvent` is returned. When the workflow encounters a returned `StopEvent`, it immediately stops the workflow and returns whatever the result was.
+
+In this case, the result is a string, but it could be a dictionary, list, or any other object.
+
+### Running the Workflow
+
+```python
+w = JokeFlow(timeout=60, verbose=False)
+result = await w.run(topic="pirates")
```

Comment:
This being async-first will likely trip up a few people. It might be worth doing the (painful) thing of having sync and async entry points